// react-context-mutex/dist/react-context-mutex.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);

var MutexContext = React__default.createContext([]);

var createMutex = function (context) {
    return /** @class */ (function () {
        function Mutex(id) {
            var _this = this;
            this.run = function (resolveCallback, rejectCallback) {
                if (!_this.isLocked()) {
                    resolveCallback();
                }
                else {
                    if (rejectCallback && typeof rejectCallback === 'function') {
                        rejectCallback();
                    }
                }
            };
            this.lock = function () {
                if (!_this.isLocked()) {
                    context.push(_this.id);
                }
            };
            this.unlock = function () {
                var index = context.indexOf(_this.id);
                if (index > -1) {
                    context.splice(index, 1);
                }
            };
            this.isLocked = function () {
                return context.indexOf(_this.id) > -1;
            };
            this.id = id;
        }
        return Mutex;
    }());
};

var useMutex = function () {
    var mutexContext = React.useContext(MutexContext);
    var MutexRunner = createMutex(mutexContext);
    return MutexRunner;
};

exports.useMutex = useMutex;
